### PR TITLE
Added test to validate that usage is displayed by admin tool for backup

### PR DESCRIPTION
**Pull request by phughk**

### CODE REVIEW

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Test
+    public void outputMatchesExpectedForMissingBackupDir() throws UnsupportedEncodingException
+    {
+        // when
+        String output = runBackup();
+
+        // then
+        String reason = "Missing argument 'backup-dir'\n\n";
+        assertEquals( reason + BackupHelpOutput.BACKUP_OUTPUT, output );
```
Comment (andrewkerr9000) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: These are going to be very brittle. Maybe use Matchers.containsString to sanity check the output, without breaking on every whitespace change.

**Review by andrewkerr9000 — COMMENTED** [review]
I suggest leaving the tests less brittle.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java (hunk 1 above)
Comment (praveenag) [reply enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: I disagree. This is probably the only time we test usage output and we should avoid random whitespaces anyways. So if this test breaks on whitespace, it's a good thing -> we can catch it, fix it.

**Review by praveenag — APPROVED** [review]

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java (hunk 1 above)
Comment (lutovich) [reply enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: I agree with @andrewkerr9000. There is now need to be this strict. It's a responsibility of the command line framework to print help correctly. Here we are just interested in presence of "Missing argument..." string. Class `BackupHelpOutput` is also not very pretty and imho should be removed.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        // then
+        String reason = "Missing argument 'name'\n\n";
+        assertEquals( reason + BackupHelpOutput.BACKUP_OUTPUT, output );
+    }
+
+    @Test
+    public void incorrectBackupDirectory() throws IOException
+    {
+        // when
+        Path backupDirectoryResolved = new File( "." ).toPath().toRealPath().resolve( "non_existing_dir" );
```
Comment (lutovich) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: Why `new File( "." ).toPath()` and not `Paths.get(".")`?
Same in `#runBackup()` method.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        Path configDir = new File( "." ).toPath();
+        List<String> params = new ArrayList();
+        params.add( "backup" );
+        params.addAll( Arrays.asList( args ) );
+        String[] argArray = params.toArray( new String[params.size()] );
+        subject.execute( homeDir, configDir, argArray );
+        return baos.toString();
+    }
+
+    private ParametrisedOutsideWorld capturableOutputOutsideWorld( PrintStream printStream )
```
Comment (lutovich) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: There exists a JUnit rule that captures `System.out` - `SuppressOutput`. Can it be used instead?

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java (hunk 1 above)
Comment (praveenag) [reply enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: I have specifically used test like this when I was refactoring in the past. The best part about the test was it was brittle -> led me to fix things and move on. W/o this test the only way I can figure out changes to the help text after a refactoring was by building neo4j. Please dont change it to contains check.

Comment (phughk) [reply enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: I will split this up into 2 tests; 1 that is checking for missing argument that doesnt include the other description and another that is checking for contains on full help.

File: enterprise/backup/src/main/java/org/neo4j/backup/impl/OnlineBackupContextBuilder.java [hunk 4]
```diff
@@ -191,12 +191,25 @@ public OnlineBackupContext createContext( String... args ) throws IncorrectUsage
...
+    private Path getFolder( Arguments arguments ) throws CommandFailed
+    {
+        Path path = arguments.getMandatoryPath( ARG_NAME_BACKUP_DIRECTORY );
+        try
+        {
+            return path.toRealPath();
+        }
+        catch ( IOException e )
+        {
+            throw new CommandFailed( String.format( "Directory '%s' does not exist.", path.toFile().toString() ) );
```
Comment (chrisvest) [inline enterprise/backup/src/main/java/org/neo4j/backup/impl/OnlineBackupContextBuilder.java]: The `toString` from `Path` should do the same, and will be implicitly called on any object processed by `%s`.

File: enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java [hunk 5]
```diff
@@ -32,7 +32,7 @@
 /**
  * An outside world where you can pick and choose which input/output are dummies.
  */
-class ParametrisedOutsideWorld implements OutsideWorld
+public class ParametrisedOutsideWorld implements OutsideWorld
```
Comment (chrisvest) [inline enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java]: Should this be in test sources?

File: enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupHelpOutput.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+public class BackupHelpOutput
+{
+    public static final String BACKUP_OUTPUT = backupOutput();
+
+    private static String backupOutput()
+    {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        stringBuilder.append( "usage: neo4j-admin backup --backup-dir=<backup-path> --name=<graph.db-backup>\n" );
```
Comment (chrisvest) [inline enterprise/backup/src/test/java/org/neo4j/backup/impl/BackupHelpOutput.java]: All these `\n`, don't they make this platform dependent?

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Test
+    public void incorrectBackupDirectory() throws IOException
+    {
+        // when
+        Path backupDirectoryResolved = new File( "." ).toPath().toRealPath().resolve( "non_existing_dir" );
+        String output = runBackup( "--backup-dir=non_existing_dir", "--name=mybackup" );
+
+        // then
+        String reason = String.format( "command failed: Directory '%s' does not exist.\n", backupDirectoryResolved.toFile().toString() );
```
Comment (chrisvest) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/BackupUsageTest.java]: Again, I don't think there's a need for the `.toFile().toString()` bit.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedBlockerLocator.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+class AugmentedBlockerLocator
+{
+    static BlockerLocator fromList( AdminCommand.Blocker... blockers )
+    {
+        return commandName -> Arrays.stream( blockers ).filter( blocker -> blocker.commands().contains( commandName ) ).collect( Collectors.toList() );
+    }
```
Comment (chrisvest) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedBlockerLocator.java]: This is only used in one place, where we don't even give it any arguments. I don't think we need this class. The test can just pass `ArrayList::new` or something like that.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedCommandLocator.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+class AugmentedCommandLocator
+{
+    /**
+     * Useful for tests where you know the list of providers but don't want to use the service locator
+     *
+     * @param providers list of command providers that the command locator will use
+     * @return a service locator based command locator that uses provided list of commands
+     */
+    static CommandLocator fromFixedArray( AdminCommand.Provider... providers )
```
Comment (chrisvest) [inline enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedCommandLocator.java]: If meant for this to be used by other tests (do you have any in mind?) then maybe the `enterprise/backup` module isn't the right home for it.

File: enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java (hunk 5 above)
Comment (phughk) [reply enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java]: It should definitely have a `test` scope. It is only used in backup module tests at the moment.

File: enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedCommandLocator.java (hunk 9 above)
Comment (phughk) [reply enterprise/backup/src/test/java/org/neo4j/commandline/admin/AugmentedCommandLocator.java]: I don't have any in mind. @praveenag also suggested that I should use the service loader instead o this, should I remove this method and use that instead?

File: enterprise/backup/src/main/java/org/neo4j/backup/impl/ParameterisedOutsideWorld.java [hunk 10]
```diff
@@ -41,7 +41,7 @@
     private final InputStream stdin;
     private final FileSystemAbstraction fileSystemAbstraction;
 
-    ParametrisedOutsideWorld( Console console, OutputStream stdout, OutputStream stderr, InputStream stdin, FileSystemAbstraction fileSystemAbstraction )
+    public ParameterisedOutsideWorld( Console console, OutputStream stdout, OutputStream stderr, InputStream stdin, FileSystemAbstraction fileSystemAbstraction )
```
Comment (chrisvest) [inline enterprise/backup/src/main/java/org/neo4j/backup/impl/ParameterisedOutsideWorld.java]: Looks like this line is now too long for our checkstyle :)

File: enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java (hunk 5 above)
Comment (phughk) [reply enterprise/backup/src/main/java/org/neo4j/backup/impl/ParametrisedOutsideWorld.java]: Resovled -> moved to `command-line` module `src/main` and kept `public`

**Pull request merged**